src/abi/abi_type.ts

```diff
@@ -0,0 +1,846 @@
+      if (arrayLength > MAX_LEN) {
+        throw new Error(`array length exceeds limit ${MAX_LEN}`);
+      }
+      // Parse the array element type
+      const arrayType = Type.Of(stringMatches[1]);
+      return new ArrayStaticType(arrayType, arrayLength);
+    }
+    if (str.startsWith('uint')) {
+      // Checks if the parsed number contains only digits, no whitespaces
+      const digitsOnly = (string) =>
+        [...string].every((c) => '0123456789'.includes(c));
+      const typeSizeStr = str.slice(4, str.length);
+      if (!digitsOnly(typeSizeStr)) {
+        throw new Error(`malformed uint string: ${typeSizeStr}`);
+      }
+      const typeSize = parseInt(typeSizeStr, 10);
+      if (typeSize > MAX_LEN) {
+        throw new Error(`malformed uint string: ${typeSize}`);
+      }
+      return new UintType(typeSize);
+    }
+    if (str === 'byte') {
+      return new ByteType();
+    }
+    if (str.startsWith('ufixed')) {
+      const stringMatches = str.match(ufixedRegexp);
+      if (stringMatches.length !== 3) {
+        throw new Error(`malformed ufixed type: ${str}`);
+      }
+      const ufixedSize = parseInt(stringMatches[1], 10);
+      const ufixedPrecision = parseInt(stringMatches[2], 10);
+      return new UfixedType(ufixedSize, ufixedPrecision);
+    }
+    if (str === 'bool') {
+      return new BoolType();
+    }
+    if (str === 'address') {
+      return new AddressType();
+    }
+    if (str === 'string') {
+      return new StringType();
+    }
+    if (str.length >= 2 && str[0] === '(' && str[str.length - 1] === ')') {
+      const tupleContent = TupleType.parseTupleContent(
+        str.slice(1, str.length - 1)
+      );
+      if (tupleContent.length === 0) {
+        return new TupleType([]);
+      }
+      const tupleTypes: Type[] = [];
+      for (let i = 0; i < tupleContent.length; i++) {
+        const ti = Type.Of(tupleContent[i]);
+        tupleTypes.push(ti);
+      }
+      return new TupleType(tupleTypes);
+    }
+    throw new Error(`cannot convert a string ${str} to an ABI type`);
+  }
+}
+
+export class UintType implements Type {
+  bitSize: number;
+
+  constructor(size: number) {
+    if (size % 8 !== 0 || size < 8 || size > 512) {
+      throw new Error(`unsupported uint type bitSize: ${size}`);
+    }
+    this.bitSize = size;
+  }
+
+  toString() {
+    return `uint${this.bitSize}`;
+  }
+
+  equal(other: UintType) {
+    return (
+      this.constructor === other.constructor && this.bitSize === other.bitSize
+    );
+  }
+
+  IsDynamic() {
+    return false;
+  }
+
+  ByteLen() {
+    return this.bitSize / 8;
+  }
+
+  Encode(value: bigint | number) {
+    if (
+      (typeof value !== 'bigint' && typeof value !== 'number') ||
+      value >= BigInt(2 ** this.bitSize) ||
+      value < 0n
+    ) {
+      throw new Error(
+        `${value} is not a non-negative int or too big to fit in size uint${this.bitSize}`
+      );
+    }
+    if (typeof value === 'number' && value >= 2 ** 53) {
```

Comment:
You should use `Number.isSafeInteger(value)` instead. As a bonus, this will also error on decimal numbers